src/fabric.c

```diff
@@ -437,42 +437,79 @@ static struct fi_provider *ofi_get_hook(const char *name)
  */
 static void ofi_ordered_provs_init(void)
 {
-	char *ordered_prov_names[] = {
-		"efa", "psm2", "opx", "usnic", "gni", "bgq", "verbs",
-		"netdir", "psm3", "ucx", "ofi_rxm", "ofi_rxd", "shm",
-
+	struct load_entry ordered_load_list[] = {
+		DEFINE_LOAD_ENTRY("efa", (HAVE_EFA | HAVE_EFA_DL)),
+		DEFINE_LOAD_ENTRY("psm2", (HAVE_PSM2 | HAVE_PSM2_DL)),
+		DEFINE_LOAD_ENTRY("opx", (HAVE_OPX | HAVE_OPX_DL)),
+		DEFINE_LOAD_ENTRY("usnic", (HAVE_USNIC | HAVE_USNIC_DL)),
+		DEFINE_LOAD_ENTRY("gni", (HAVE_GNI | HAVE_GNI_DL)),
+		DEFINE_LOAD_ENTRY("bgq", (HAVE_BGQ | HAVE_BGQ_DL)),
+		DEFINE_LOAD_ENTRY("verbs", (HAVE_VERBS | HAVE_VERBS_DL)),
+		DEFINE_LOAD_ENTRY("netdir", 0),
+		DEFINE_LOAD_ENTRY("psm3", (HAVE_PSM3 | HAVE_PSM3_DL)),
+		DEFINE_LOAD_ENTRY("ucx", (HAVE_UCX | HAVE_UCX_DL)),
+		DEFINE_LOAD_ENTRY("ofi_rxm", (HAVE_RXM | HAVE_RXM_DL)),
+		DEFINE_LOAD_ENTRY("ofi_rxd", (HAVE_RXD | HAVE_RXD_DL)),
+		DEFINE_LOAD_ENTRY("shm", (HAVE_SHM | HAVE_SHM_DL)),
 		/* Initialize the socket based providers last of the
 		 * standard providers.  This will result in them being
 		 * the least preferred providers.
 		 */
 
 		/* Before you add ANYTHING here, read the comment above!!! */
-		"udp", "tcp", "sockets", "net", /* NOTHING GOES HERE! */
+		DEFINE_LOAD_ENTRY("udp", (HAVE_UDP | HAVE_UDP_DL)),
+		DEFINE_LOAD_ENTRY("tcp", (HAVE_TCP | HAVE_TCP_DL)),
+		DEFINE_LOAD_ENTRY("sockets", (HAVE_SOCKETS | HAVE_SOCKETS_DL)),
+		DEFINE_LOAD_ENTRY("net", 0), /* NOTHING GOES HERE! */
 		/* Seriously, read it! */
 
 		/* These are hooking providers only.  Their order
 		 * doesn't matter
 		 */
-		"ofi_hook_perf", "ofi_hook_trace", "ofi_hook_profile", "ofi_hook_debug",
-		"ofi_hook_noop", "ofi_hook_hmem", "ofi_hook_dmabuf_peer_mem",
+                DEFINE_LOAD_ENTRY("ofi_hook_perf", (HAVE_PERF | HAVE_PERF_DL)),
+                DEFINE_LOAD_ENTRY("ofi_hook_trace", (HAVE_TRACE | HAVE_TRACE_DL)),
+                DEFINE_LOAD_ENTRY("ofi_hook_profile", (HAVE_PROFILE_DL)),
+                DEFINE_LOAD_ENTRY("ofi_hook_debug", (HAVE_HOOK_DEBUG_DL)),
+                DEFINE_LOAD_ENTRY("ofi_hook_noop", 0),
+                DEFINE_LOAD_ENTRY("ofi_hook_hmem", (HAVE_HOOK_HMEM_DL)),
+                DEFINE_LOAD_ENTRY("ofi_hook_dmabuf_peer_mem",
+                        (HAVE_DMABUF_PEER_MEM | HAVE_DMABUF_PEER_MEM_DL)),
 
 		/* So do the offload providers. */
-		"off_coll",
+		DEFINE_LOAD_ENTRY("off_coll", 0),
 	};
 	struct ofi_prov *prov;
 	int num_provs, i;
 
-	num_provs = sizeof(ordered_prov_names) / sizeof(ordered_prov_names[0]);
+	num_provs = sizeof(ordered_load_list) / sizeof(ordered_load_list[0]);
 
 	for (i = 0; i < num_provs; i++) {
-		prov = ofi_alloc_prov(ordered_prov_names[i]);
+		/* If restricted DL is enabled, then only attempt to load
+		 * providers which were present at compile-time that were
+		 * enabled as 'yes' or 'dl'
+		 */
+		if (HAVE_RESTRICTED_DL && !ordered_load_list[i].should_load)
+			continue;
+
+		prov = ofi_alloc_prov(ordered_load_list[i].provider_name);
```

Comment:
This won't be able to limit which providers to load.

As I commented on #9296, this only allocate a provider structure so that they appear in the predefined order. Providers not allocated here are still going to be searched and opened later.  

See `ofi_init_dir`, `ofi_reg_dl_prov` and `ofi_register_provider` for more details.

Reply:
I've reviewed those functions before, but I don't see what you are trying to point out. Please help me to see what you are trying to point out. 

> See ofi_init_dir, ofi_reg_dl_prov and ofi_register_provider for more details.

`ofi_ini_dir` is only used by libfabric when `FI_PROVIDER_PATH` is provided. It happens here: https://github.com/ofiwg/libfabric/blob/main/src/fabric.c#L760. `dirs` is only set if provdir is set, which is directly associated with `FI_PROVIDER_PATH` on line 744. If `FI_PROVIDER_PATH` isn't defined, then libfabric will search the pre-populated `prov_head`. The `prov_head` is populated as part of `fi_ini` through `ofi_ordered_provs_init`. 

There are two entry points for `prov_head`/`ofi_insert_prov`: `ofi_register_provider`, and `ofi_ordered_provs_init`. For any other provider to be added to the list, it would need to go through the `ofi_register_provider` path -> which has entry points from `ofi_reg_dl_prov`, and `fi_ini`. 

When libfabric traverses into `ofi_register_provider` from `ofi_reg_dl_prov`, it only happens from a limited set of contexts: 
- when a provider is being searched via `FI_PROVIDER_PATH` (`ofi_ini_dir`-> `ofi_reg_dl_prov` -> `ofi_register_provider` -> `ofi_insert_prov`)
- when a provider already exists within the `prov_head` list, and DL providers in the process of being discovered by libfabric `ofi_find_prov_libs`. At line 700, the contents of `prov_head` is what was populated by `ofi_ordered_provs_init` (`ofi_find_prov_libs` -> `ofi_reg_dl_prov` -> `ofi_register_provider` -> `ofi_insert_prov`)

Otherwise, when `ofi_register_provider` is called from `fi_ini`, it is done within the context of basic initialization after `ofi_load_dl_prov` is called on line 888.  Since `ofi_load_dl_prov` is called before the [static initializers of the providers](https://github.com/ofiwg/libfabric/blob/main/src/fabric.c#L890-L919), none of the static initializers have the opportunity to affect the `prov_head` list to determine what providers could be discovered as DL loadable libraries. Even if the static initializers were moved ahead if the call to `ofi_load_dl_prov`, the first parameter of `ofi_register_provider` is the `<prov>_INIT` define, which would be NULL if support for the library wasn't present at the time libfabric was compiled. To me, this means that unless the provider is in `ofi_orders_provs_init`, it won't have DL provider support as a separate module shipped by the vendor independent of the distribution provider (libfabric from SuSE with libsm2-fi.so from the vendor -- as an example, if SuSE didn't enable sm2 in libfabric via --enable-sm2=yes|dl). As each of the static initializers are called, `fi_register_provider(GNI_INIT, NULL)` as an example, it will initialize the core provider that was built in, but will not attempt to load any DL providers because there isn't a call at this point which feeds back into the DL provider loads. Libfabric will add the pre-built/in-library version of the provide via the call to `ofi_insert_prov` and `ofi_init_prov` on lines 559-560, but it stops there. There are no additional calls to `ofi_reg_dl_prov` after the static initializers are called. Providers that do not exist in `prov_head` prior to the last call of `ofi_reg_dl_prov` are not discoverable. 

So at lines [744-747](https://github.com/ofiwg/libfabric/blob/main/src/fabric.c#L745-L747), if `FI_PROVIDER_PATH` isn't provided, then it will use what information was populated already to find dl libraries. `dirs` will be populated with the OFI `PROVDLDIR` default, and that will be the only path searched by `ofi_ini_dir`. 

Otherwise, if a value is provided to `FI_PROVIDER_PATH`, libfabric will delimit the string and traverse it by order given in the string, with special treatment for paths prefixed with `@`. 

The real problem that is being addressed here is the behavior for https://github.com/ofiwg/libfabric/blob/main/src/fabric.c#L746, as it searches the many different possible providers that could exist that do not. Every entry on `prov_head` will result in a call to `dlopen`. The only way that it does not happen is if `FI_PROVIDER_PATH` is set. 

-----------------------

This is observable from a fairly simple example. 

```
./autogen.sh 
./configure --enable-only --enable-tcp
./util/fi_info -p udp 
```
 
 No providers will be shown because only the TCP provider has been compiled in. However, every provider within the `ordered_prov_names` list will be called via `dlopen` on the expected DL provider name `lib<prov>-fi.so`. This results in N(number of providers known) * M (number of directories in LD_LIBRARY_PATH) * P (number of ranks per node) * Q (number of nodes allocated to the job) failed queries in the worst case scenario. This is a realistic problem for deployments which only use a very small subset of libfabric providers, and only gets worse at scale. 
 
 If you add debug code at line 645 to change the FI_DBG to a printf (required because it doesn't seem like the logging subsystem is fully set up by this time in the code), then you will see all of the spurious calls to dlopen, and if they don't exist on the system -- then every probe made by libdl will fail. The list of directories is longer than indicated by LD_LIBRARY_PATH because libdl does search extra directories for architecture specific variants of libraries. @timattox caught that.

Reply:
What I tried to point out was that the method wouldn't prevent a provider from being loaded even if it is not in the list. However, you clarification indicates that this is not the purpose.

> The real problem that is being addressed here is the behavior for https://github.com/ofiwg/libfabric/blob/main/src/fabric.c#L746, as it searches the many different possible providers that could exist that do not. Every entry on prov_head will result in a call to dlopen.

If that's all what you want to address, then I think it's fine.

Nitpick:

> ofi_ini_dir is only used by libfabric when FI_PROVIDER_PATH is provided

When FI_PROVIDER_PATH is not set, this function is still called with a compile time set path. I see you also mentioned that later.